### ISSUE TITLE
A library developer should be able to select matching names given a list

### ISSUE THREAD

**Issue by wdanilo** [enso-org/enso#4173]
This task is automatically imported from the old Task Issue Board and it was originally created by James Dunkerley.
Original issue is [here](https://www.pivotaltracker.com/story/show/180855277).
<hr />

## Team
Technical Designer: James Dunkerley
Implementator: Radosław Waśko
Acceptance/QA: James Dunkerley

## Why

**As** a library developer
**I want** to be able to select a subset of a text vector, based on some match criteria and a second vector
**So that** I can work with the matching subset

## Acceptance Criteria

```gherkin
Scenario: 
GIVEN I'm implementing an operation manipulating columns of a Table
WHEN I provide criteria for selecting columns to modify
THEN the helper function returns columns matching that criteria so that my operation can transform them in an intended way
```

## Notes
The more high-level function should have the following interface:
```
match : Vector Text -> Vector Text -> Matching_Strategy -> Missing_Behavior -> Vector Text ! Column_Missing
match <full_list> <values_to_match> <matching_strategy> <missing_behavior>
```

`values_to_match` can be just a column (then it is selected as-is if it is also present in `full_list`) or textual criteria which are then handled according to the `matching_strategy`.

The `matching_strategy` is described by the following type:
```
type Matching_Strategy
   type Exact case_sensitive=True # a name is matched if its exact value is present in `values_to_match`
   type Regex case_sensitive=True  # a name is matched if its value matches any regex present in `values_to_match`
```

*In the future we may also want to extend it with `Wildcard` and `SQL_Like` for matching expressions like `*Foo*` or `%Foo%`, but it is not in the scope of this task.*

The `missing_behavior` specifies if a warning or error should be reported if no columns match an entry in `columns_to_match`. It is out of scope for this task to implement the warning response.

To implement the above function we also should have a helper function which just checks if a single given name matches a single matching criterion. It will allow library developers to easily extend this mechanism if they need more specific information than just a list of names which match the criteria. Depending on the complexity, it may be worth to implement the underlying function in Java for speed, if necessary.

### Additional considerations 

The implementation should favour simplicity over maximizing performance - O(N^2) is acceptable since it will be reasonably fast even for 1000s of components - the bottleneck is actually processing the data, not selecting the columns. If it turns out we need more optimized implementations, we can upgrade them when necessary. 

The implementation should be universal in that the helper function should be usable by both InMemory and DB and other Table implementations (so it should be agnostic of the underlying Table implementation).

## Blockers:
[#180899205](https://www.pivotaltracker.com/story/show/#180899205) resolved

## Comments:
**Radosław Waśko** reports a new **STANDUP** for today (2022-01-10):

**Today:** Again reviewing the design, clarifying the next steps with James. The long team meeting. Starting work on a column matching utility to be used internally by SELECT-related operations. Writing down some further considerations for the design that may need to be discussed before this utility is finished. It should be finished by 2022-01-12.

**Next Day:** Next day I will be working on the [#same task](https://www.pivotaltracker.com/story/show/180855277). Refining some design stuff, hopefully finally getting the final task split for the SELECT story and continuing work on the match utility.

**Additional Information:** It may be finished earlier, but there are still design decisions to discuss so I'm aiming for slightly later. (Enso Bot - Jan 10, 2022)
<hr />
**Radosław Waśko** reports a new **STANDUP** for yesterday (2022-01-11):

**Today:** Discussing some principles we want to keep in our implementations with James, was working on the match function and found some edge cases that we did not consider earlier. An initial implementation of single matching with some tests is done, but tests still need to be fixed. It should be finished by 2022-01-12.

**Next Day:** Next day I will be working on the [#same task](https://www.pivotaltracker.com/story/show/180855277). Fix the tests, add multi-matching and test it. (Enso Bot - Jan 12, 2022)
<hr />
**Radosław Waśko** reports a new **🔴 DELAY** for yesterday (2022-01-12):

**Summary:** There is **2 days** delay in implementation of the [A library developer should be able to select matching names given a list (#180855277) task](https://www.pivotaltracker.com/story/show/180855277).
It **will cause 0 days delay** for the delivery of this sprint.

**Delay Cause:** Primary issue is that it turned out we need a `Vector.distinct` operation to implement the functionality, but stdlib does not provide it - so I need to switch tasks to implement the missing operation on Vector. I then need another day to update docs and merge the original task. I also slightly underestimated the complexity of multiple matching and the overhead of having quite a few meetings on each day. (Enso Bot - Jan 13, 2022)
<hr />
**Radosław Waśko** reports a new **STANDUP** for yesterday (2022-01-12):

**Today:** Most of the implementation and tests has been done. It is blocked by `Vector.distinct`. It should be finished by 2022-01-14.

**Next Day:** Next day I will be working on the [##180899205 task](https://www.pivotaltracker.com/story/show/180899205). I need to switch over to implement `Vector.distinct` as it is blocking matching. (Enso Bot - Jan 13, 2022)
<hr />
**Radosław Waśko** reports a new **🔴 DELAY** for yesterday (2022-01-14):

**Summary:** There is **3 days** delay in implementation of the [A library developer should be able to select matching names given a list (#180855277) task](https://www.pivotaltracker.com/story/show/180855277).
It **will cause 0 days delay** for the delivery of this sprint.

**Delay Cause:** The Vector.distinct was still not merged (CI fails (mostly on Windows) and I cannot override it myself). The design discussions took a significant amount of time, so I was able to finish he docs a bit later and will be able to get the PR reviewed on Monday. (Enso Bot - Jan 15, 2022)
<hr />
**Radosław Waśko** reports a new **STANDUP** for yesterday (2022-01-14):

**Today:** Back to the matching task. Design reviews and discussions about the Select library, fixing the syntax. Adding docs and updates after the call to the matching task. A write-up with some questions about Atom typing which will help us design the interfaces. It should be finished by 2022-01-17.

**Next Day:** Next day I will be working on the [##180855277 task](https://www.pivotaltracker.com/story/show/180855277). Getting the matching task reviewed and merged. Planning next tasks for the following week. (Enso Bot - Jan 15, 2022)
<hr />
**Radosław Waśko** reports a new **🔴 DELAY** for yesterday (2022-01-16):

**Summary:** There is **1 day** delay in implementation of the [A library developer should be able to select matching names given a list (#180855277) task](https://www.pivotaltracker.com/story/show/180855277).
It **will cause 0 days delay** for the delivery of this sprint.

**Delay Cause:** When designing the actual select_column implementation we figured out match can be slightly extended so that it can be used in a cleaner way. It's waiting on CI and for QA for final delivery. (Enso Bot - Jan 17, 2022)
<hr />
**Radosław Waśko** reports a new **STANDUP** for today (2022-01-17):

**Today:** Final touches to the matching task, extended it with ability to work with not only names but named complex object so that it can be used in a cleaner way. Added some more tests. The PR waits for a merge and QA. It should be finished by 2022-01-18.

**Next Day:** Next day I will be working on the [##180794249 task](https://www.pivotaltracker.com/story/show/180794249). Putting the pieces together to work on the select_columns implementation. (Enso Bot - Jan 17, 2022)
<hr />
**Radosław Waśko** reports a new **STANDUP** for yesterday (2022-01-18):

**Today:** Responding to some feedback on the match PR; discussing edge-cases of Vector.distinct with James and preparing a Demo for the weekly meeting (which revealed a few unrelated bugs that need reporting); also a call on CI with Michał. It should be finished by 2022-01-18.

**Next Day:** Next day I will be working on the [##180794249 task](https://www.pivotaltracker.com/story/show/180794249). Finally working on the select_columns operation (yesterday other stuff precluded this) and reporting the issues discovered yesterday. (Enso Bot - Jan 19, 2022)
<hr />
**Marcin Kostrzewa** reports a new **🔴 DELAY** for yesterday (2022-01-19):

**Summary:** There is **1 day** delay in implementation of the [A library developer should be able to select matching names given a list (#180855277) task](https://www.pivotaltracker.com/story/show/180855277).
It **will cause 1 day delay** for the delivery of this sprint.

no idea if this is a delay, I'm just trying to track time for a discussion I was pulled into, I don't have much to do with this task

**Delay Cause:** no idea if this is a delay, I'm just trying to track time for a discussion I was pulled into, I don't have much to do with this task

**Possible solutions:** let me track time without being forced to do things like this 🙂 (Enso Bot - Jan 20, 2022)
<hr />
**Marcin Kostrzewa** reports a new **STANDUP** for yesterday (2022-01-19):

**Today:** Participated in a discussion regarding dataflow error handling in the PR. It should be finished by 2022-01-19.

**Next Day:** Next day I will be working on the [##180956811 task](https://www.pivotaltracker.com/story/show/180956811). Back to types (Enso Bot - Jan 20, 2022)
<hr />
